src/inventory/slot_mapping.rs

```diff
@@ -0,0 +1,196 @@
+use crate::inventory::{Item, Slot};
+use crate::render::hud::Hud;
+use crate::render::inventory::InventoryWindow;
+use crate::render::Renderer;
+use crate::ui::{Container, VAttach};
+use std::collections::HashMap;
+use std::sync::Arc;
+
+use parking_lot::RwLock;
+
+/// A mapping of each slot and it's position on the screen. This makes the job
+/// of rendering and locating slots easier for inventories. This class only
+/// expects positions in pixels relative to the top-left corner of the window.
+pub struct SlotMapping {
+    /// A mapping of slot_id to (position, slot). The position is the distance
+    /// between the window's top-left corner, and the slot's top-left corner in
+    /// pixels. This doesn't include the border frame around each slot, just
+    /// the 16x16 content.
+    slots: HashMap<u16, ((i32, i32), Slot)>,
+    /// The size of the window in pixels.
+    size: (i32, i32),
+    dirty: bool,
+    /// A child slot mapping to be used as a fallback if a slot wasn't found
+    /// in the parent mapping. This is used to handle the base inventory.
+    child: Option<Arc<RwLock<SlotMapping>>>,
+    /// The offset between the top-left corner of the current mapping window,
+    /// and the corner of the sub-section that the child mapping manages, in
+    /// pixels.
+    child_offset: Option<(i32, i32)>,
+    /// The slot ids to be used on the parent interface to access slots in the
+    /// child. We use the index of each parent slot id as the child slot id.
+    /// For example, if set to `[5, 6, 7]`, then accessing slot id `6` on the
+    /// parent will look up the slot_id `1` on the child.
+    child_range: Option<Vec<u16>>,
+}
+
+impl SlotMapping {
+    /// Create a new slot mapping for a window with the given size in pixels.
+    pub fn new(size: (i32, i32)) -> Self {
+        Self {
+            slots: HashMap::new(),
+            size,
+            dirty: false,
+            child: None,
+            child_offset: None,
+            child_range: None,
+        }
+    }
+
+    /// Get the number of slots in this mapping, including any child slots.
+    pub fn size(&self) -> u16 {
+        (match &self.child_range {
+            Some(child_range) => self.slots.len() + child_range.len(),
+            None => self.slots.len(),
+        }) as u16
+    }
+
+    /// Set the child slot mapping. This will replace any child previously set.
+    pub fn set_child(
+        &mut self,
+        child: Arc<RwLock<SlotMapping>>,
+        offset: (i32, i32),
+        range: Vec<u16>,
+    ) {
+        self.child = Some(child);
```

Comment:
alternatively, to the alternative representations i proposed above we could probably also simply use a single struct wrapped inside an ArcSwapOption that contains `child`, `child_offset` and `child_range`

Reply:
interior mut